docs/proposals/node-local-services.md

```diff
@@ -0,0 +1,269 @@
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Node-local services](#node-local-services)
+  - [Objective](#objective)
+  - [Background](#background)
+  - [Detailed discussion](#detailed-discussion)
+    - [API changes](#api-changes)
+    - [`kube-proxy` changes](#kube-proxy-changes)
+      - [iptables kernel mode](#iptables-kernel-mode)
+      - [ipvs kernel mode](#ipvs-kernel-mode)
+      - [proxy userland mode](#proxy-userland-mode)
+    - [n-to-1 forwarding](#n-to-1-forwarding)
+  - [Implementation plan](#implementation-plan)
+    - [API](#api)
+    - [kube-proxy](#kube-proxy)
+    - [Documentation](#documentation)
+  - [Future work](#future-work)
+  - [Security considerations](#security-considerations)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# Node-local services
+
+Author: @therc
+
+Date: July 2016
+
+Status: Design in progress
+
+## Objective
+
+Users should be able to create a service that connects pods to a daemon that is
+running on the same node (and not elsewhere). Possible use cases:
+
+ - monitoring agents and forwarders such as datadog and
+   [tallier](https://github.com/reddit/tallier)
+ - connection poolers and service discovery systems such as
+   [pgbouncer](https://pgbouncer.github.io/) and
+   [synapse](https://github.com/airbnb/synapse)
+ - logging agents such as fluentd
+ - authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy),
+   [kube2iam](https://github.com/jtblin/kube2iam) or loasd ([#2209](https://github.com/kubernetes/kubernetes/issues/2209))
+
+At the very least, this new kind of service should support 1:1
+_service_->_daemon_ mappings. For cases like `loasd`, it would be ideal to
+support n:1 mapping: a number of services all getting forwarded to the same
+daemon, with the daemon able to tell which service each connection was meant
+for.
+
+## Background
+
+Thanks to DaemonSets, it is possible to schedule daemon pods on every node (or a
+specific subset if node labels are specified). Local discovery is more
+complicated and currently requires e.g. iptables hacks that bypass Kubernetes
+altogether. This is because existing types of services will pick a random
+service pod when a client establishes a new connection. More often than not, the
+client pod will talk to a service pod running on another machine.
+
+This is not acceptable for some use cases. The daemon might return node-specific
+information downstream or report it upstream to a server. If it were to receive
+traffic from another machine, data returned or forwarded would be inaccurately
+tagged. In other cases, the main concern is reducing cross-node traffic, be it
+for bandwidth or security reasons.
+
+Thus, a new service type is needed that always points to the node-local daemon
+pod.
+
+## Detailed discussion
+
+Node-local services can reuse most of the existing plumbing.
```

Comment:
I apologize if this gets a little long.  I'm trying to not fall off the rails here, but I think this proposal can be more (alarm bells should be ringing).

What does "local" mean?  Same-node is great, and probably covers 75% or more of use-cases.  But we have definitely been asked for same-rack, or same-enclosure, or same zone Services.  So what does "local" mean? It means "same something as me".

Superficially this feels the same as the `externalTraffic=OnlyLocal` (@bprashanth), except that is explicitly same-node and explicitly external traffic.  I think the use of affinity here is clever, and if at the end of this I have talked myself in a circle (46% odds) maybe that's the right approach.  But if we want to allow more arbitrary locality, it's not parameterized, and there are some orthogonality concerns.

Consider as an alternative, a new field (obviously, we would trial this as an annotation or whatever we have ultimately decided to do there) called `locality`, whose value is a string label key.

```
    locality: "kubernetes.io/hostname"
```

When choosing a backend for a new client we would do something like:

```
read my own Node's label value for the locality key
only consider Pods on Nodes whose label value for the locality key is the same
```

Conveniently, Endpoints already carry the NodeName.  Watching Nodes and tracking labels would not be difficult.

Result:
   * If I run my pod as a daemonset, I can use hostname as locality.
   * If I run my pod as a per-rack pods (e.g. nodeSelector for rack=X), then I can use my custom rack label.
   * If I run my pod as a per-zone pods, then I can use the failure-domain.../zone label.  I can even run multiple replicas per-zone, and locality rules allow me to choose any backend in my same zone.  This says to me that affinity is orthogonal.

What do you think?

This has some overlap with federation, a little ( @nikhiljindal )

This has some overlap with service gateways ( @enricoschiattarella )

This has some overlap with scheduling ( @davidopp )

@bgrant0607 because API idea

@kubernetes/sig-network-misc 
@matchstick